x-pack/plugins/apm/server/saved_objects/migrations/migrate_legacy_apm_indices_to_space_aware.ts

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { CoreStart } from 'src/core/server';
+import {
+  APM_INDICES_SAVED_OBJECT_ID,
+  APM_INDICES_SAVED_OBJECT_TYPE,
+  APM_INDICES_SPACE_SAVED_OBJECT_ID,
+  APM_INDICES_SPACE_SAVED_OBJECT_TYPE,
+} from '../../../common/apm_saved_object_constants';
+import { ApmIndicesConfig } from '../../routes/settings/apm_indices/get_apm_indices';
+
+export async function migrateLegacyAPMIndicesToSpaceAware({
+  coreStart,
+}: {
+  coreStart: CoreStart;
+}) {
+  const repository = coreStart.savedObjects.createInternalRepository(['space']);
+  try {
+    // Fetch legacy APM indices
+    const legacyAPMIndices = await repository.get<Partial<ApmIndicesConfig>>(
+      APM_INDICES_SAVED_OBJECT_TYPE,
+      APM_INDICES_SAVED_OBJECT_ID
+    );
+
+    // Fetch spaces available
+    const spaces = await repository.find({
+      type: 'space',
+      page: 1,
+      perPage: 10_000, // max number of spaces as of 8.1
+      fields: ['name'], // to avoid fetching *all* fields
+    });
+
+    // Create new APM indices space aware for all spaces available
+    await repository.bulkCreate(
+      spaces.saved_objects.map(({ id: spaceId }) => ({
+        id: APM_INDICES_SPACE_SAVED_OBJECT_ID,
+        type: APM_INDICES_SPACE_SAVED_OBJECT_TYPE,
+        initialNamespaces: [spaceId],
+        attributes: legacyAPMIndices.attributes,
+      }))
+    );
+
+    // Delete legacy APM indices
+    await repository.delete(
+      APM_INDICES_SAVED_OBJECT_TYPE,
+      APM_INDICES_SAVED_OBJECT_ID
+    );
+  } catch (e) {
+    // Does nothing when an exception happens (repository.get throws an exception when a saved object is not found)
+  }
```

Comment:
This empty catch is probably a bad idea 😅  I say this because my team just ran into a problem with an empty catch that caused an unexpected bug, and the wound is still fresh!

I'd suggest making a few changes:

1. Treat _only_ the `get` "Not Found" error as an **expected error**
2. For all **unexpected errors**, log an appropriate message
3. Keep a snapshot of the legacy APM indices object type to ensure that future changes to `ApmIndicesConfig` overlap sufficiently (the type checker will make sure of this for us!)
   - This will only matter if a user upgrades straight from 8.1 to 8.42 for example, skipping a bunch of minors... it's something we support so we should try to think about it proactively


Diff of suggested changes, not including my naming suggestions in my other comment:

```diff
diff --git a/x-pack/plugins/apm/server/saved_objects/migrations/migrate_legacy_apm_indices_to_space_aware.ts b/x-pack/plugins/apm/server/saved_objects/migrations/migrate_legacy_apm_indices_to_space_aware.ts
index a5c7401d61b..8ae691f2bc8 100644
--- a/x-pack/plugins/apm/server/saved_objects/migrations/migrate_legacy_apm_indices_to_space_aware.ts
+++ b/x-pack/plugins/apm/server/saved_objects/migrations/migrate_legacy_apm_indices_to_space_aware.ts
@@ -4,7 +4,12 @@
  * 2.0; you may not use this file except in compliance with the Elastic License
  * 2.0.
  */
-import { CoreStart } from 'src/core/server';
+import type {
+  CoreStart,
+  Logger,
+  ISavedObjectsRepository,
+} from 'src/core/server';
+import { SavedObjectsErrorHelpers } from 'src/core/server';
 import {
   APM_INDICES_SAVED_OBJECT_ID,
   APM_INDICES_SAVED_OBJECT_TYPE,
@@ -13,29 +18,62 @@ import {
 } from '../../../common/apm_saved_object_constants';
 import { ApmIndicesConfig } from '../../routes/settings/apm_indices/get_apm_indices';
 
+/**
+ * Legacy APM indices config object attributes (pre-8.2).
+ * We need to keep a snapshot of this interface to guard against any incompatible ApmIndicesConfig interface changes in the future.
+ */
+interface LegacyApmIndicesConfig {
+  sourcemap: string;
+  error: string;
+  onboarding: string;
+  span: string;
+  transaction: string;
+  metric: string;
+  apmAgentConfigurationIndex: string;
+  apmCustomLinkIndex: string;
+}
+
+async function fetchLegacyAPMIndices(repository: ISavedObjectsRepository) {
+  try {
+    // Fetch legacy APM indices
+    const legacyAPMIndices = await repository.get<
+      Partial<LegacyApmIndicesConfig>
+    >(APM_INDICES_SAVED_OBJECT_TYPE, APM_INDICES_SAVED_OBJECT_ID);
+    return legacyAPMIndices;
+  } catch (err) {
+    if (SavedObjectsErrorHelpers.isNotFoundError(err)) {
+      // This is expected after the legacy APM indices object has been migrated for the first time.
+      return null;
+    }
+    throw err;
+  }
+}
+
 export async function migrateLegacyAPMIndicesToSpaceAware({
   coreStart,
+  logger,
 }: {
   coreStart: CoreStart;
+  logger: Logger;
 }) {
   const repository = coreStart.savedObjects.createInternalRepository(['space']);
   try {
     // Fetch legacy APM indices
-    const legacyAPMIndices = await repository.get<Partial<ApmIndicesConfig>>(
-      APM_INDICES_SAVED_OBJECT_TYPE,
-      APM_INDICES_SAVED_OBJECT_ID
-    );
+    const legacyAPMIndices = await fetchLegacyAPMIndices(repository);
+    if (legacyAPMIndices === null) {
+      return;
+    }
 
     // Fetch spaces available
     const spaces = await repository.find({
       type: 'space',
       page: 1,
-      perPage: 10_000, // max number of spaces as of 8.1
+      perPage: 10_000, // max number of spaces as of 8.2
       fields: ['name'], // to avoid fetching *all* fields
     });
 
     // Create new APM indices space aware for all spaces available
-    await repository.bulkCreate(
+    await repository.bulkCreate<Partial<ApmIndicesConfig>>(
       spaces.saved_objects.map(({ id: spaceId }) => ({
         id: APM_INDICES_SPACE_SAVED_OBJECT_ID,
         type: APM_INDICES_SPACE_SAVED_OBJECT_TYPE,
@@ -50,6 +88,6 @@ export async function migrateLegacyAPMIndicesToSpaceAware({
       APM_INDICES_SAVED_OBJECT_ID
     );
   } catch (e) {
-    // Does nothing when an exception happens (repository.get throws an exception when a saved object is not found)
+    logger.error('Failed to migrate legacy APM indices object: ' + e.message);
   }
 }
```

Reply:
Thanks for the great review @jportner , working on the fixes

Reply:
done https://github.com/elastic/kibana/pull/126176/commits/83e29f911e84a80dd951f8b6b9678840ed945334